### PR TITLE
Add gauss points for hex and quad elements, fix #68

### LINKED ISSUE #68: Gauss Points Implementation
Prepare the necessary files and functions for gauss point integration scheme in MPM

- Make a new class Gauss
- Change necessary functions in `particle.h` and `node.h`

Follow this [link ](https://docs.google.com/document/d/1UUTEbbgZsPed40KOf-oLHb2kaHg8N2F9FPSdR18ca9k/edit) for further documentation.

### PR DESCRIPTION
**Pull request by ezrayst**

### CODE REVIEW

File: include/gauss.h [hunk 1]
```diff
@@ -0,0 +1,154 @@
+#ifndef MPM_PARTICLE_H_
```
Comment (kks32) [inline include/gauss.h]: Why is `gauss.h` has the Particle class?

File: CMakeLists.txt [hunk 2]
```diff
@@ -55,7 +55,7 @@ add_executable(mpm
 SET(test_src
    ${mpm_SOURCE_DIR}/tests/cell_container_test.cc
    ${mpm_SOURCE_DIR}/tests/cell_test.cc
-   ${mpm_SOURCE_DIR}/tests/quad_shapefn_test.cc
+   ${mpm_SOURCE_DIR}/tests/hex_quad_test.cc
```
Comment (kks32) [inline CMakeLists.txt]: could we call quadrature as `quadrature` instead of `quad`

File: include/hex_quad.tcc [hunk 3]
```diff
@@ -0,0 +1,217 @@
...
+template <unsigned Tdim, unsigned Tnquadratures>
+mpm::HexahedronQuadrature<Tdim, Tnquadratures>::HexahedronQuadrature()
+    : mpm::QuadratureBase<Tdim, Tnquadratures>() {
+
+  static_assert(Tdim == 3, "Invalid dimension for a 3D hexahedron element");
+  static_assert(
+      (Tnquadratures == 1) || (Tnquadratures == 8) || (Tnquadratures == 27),
+      "Invalid number of quadratures");
+
+  if (Tnquadratures == 1) {
```
Comment (kks32) [inline include/hex_quad.tcc]: Could we do partial specialisation instead?

File: include/quad_base.h [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+#define MPM_QUADRATURE_BASE_H_
+
+#include <algorithm>
+#include <iostream>
+#include <limits>
+#include <vector>
+
+#include <Eigen/Dense>
+
+namespace mpm {
```
Comment (kks32) [inline include/quad_base.h]: No need to predefine a class

File: include/quad_base.h [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+template <unsigned Tdim, unsigned Tnquadratures>
+class QuadratureBase;
+}
+
+// Quadrature base class
+//! \brief Base class for quadrature
+//! \tparam Tdim Dimension
+//! \tparam Tnquadratures number of quadratures
+template <unsigned Tdim, unsigned Tnquadratures>
+class mpm::QuadratureBase {
```
Comment (kks32) [inline include/quad_base.h]: This should be an abstract class

File: include/quad_base.h [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+  //! Destructor
+  virtual ~QuadratureBase() {}
+
+  //! Return quadrature points
+  //! \param[out] qpoints Quadrature points in local coordinates
+  Eigen::Matrix<double, Tnquadratures, Tdim> quadratures() { return qpoints_; }
+
+  //! Return weights
+  //! \param[out] weights Weights for quadrature points
+  std::vector<double> weights() { return weights_; }
```
Comment (kks32) [inline include/quad_base.h]: All vectors are Eigen vectors in the MPM code.

File: include/quad_base.h [hunk 7]
```diff
@@ -0,0 +1,48 @@
+#ifndef MPM_QUADRATURE_BASE_H_
```
Comment (kks32) [inline include/quad_base.h]: This file should be just called `quadrature.h`

**Review by kks32 — CHANGES_REQUESTED** [review]

**Comment by ezrayst** [comment]
I have not changed to partial specalization, but I am having trouble in getting the test cases to run. Will check again but it is giving seg-fault now.

**Comment by kks32** [comment]
Hi @ezrayst Thank you for these. Could you try running your test cases with `./mpmtest -s` and `./mpmtest` without `-s`, sometimes it might be because your constructor fails. Do run `gdb --args ./mpmtest -s [quadrature]` if you want to just test the quadrature. you can also do `[2D][quadrature]` which will only test 2D cases with quadrature. In gdb `run` and then do `bt` to see where it fails. I'll have a look as well on your branch.

**Comment by kks32** [comment]
Hi @ezrayst Your last commit seems to have removed / renamed files incorrectly, which results in compilation issues. Could you please check? Thanks!

File: include/gauss.h (hunk 1 above)
Comment (kks32) [reply include/gauss.h]: Is this file required? It seems to be a copy of `Particle` class

File: include/hex_quadrature.tcc [hunk 8]
```diff
@@ -0,0 +1,243 @@
...
+    weights(7) = 1.;
+
+  return weights;
+}
+
+// Getting the quadratures for Tnquadratures = 27
+template <>
+inline Eigen::MatrixXd mpm::HexahedronQuadrature<3, 27>::quadratures() {
+  Eigen::Matrix<double, 27, 3> quadratures;
+    const double Qpoint_a = std::sqrt(3. / 5.);
```
Comment (kks32) [inline include/hex_quadrature.tcc]: Hi @ezrayst could you please use a lower case for all variables, we no longer use capitalisation of `const` as `const`-ing is now very common than before. Thanks!

File: include/quadrature.h [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+// Quadrature base class
+//! \brief Base class for quadrature
+//! \tparam Tdim Dimension
+//! \tparam Tnquadratures number of quadratures
+template <unsigned Tdim, unsigned Tnquadratures>
+class QuadratureBase {
+ public:
+  //! Constructor
+  //! Assign variables to zero
+  QuadratureBase() {}
```
Comment (kks32) [inline include/quadrature.h]: Set this to default.

**Review by kks32 — CHANGES_REQUESTED** [review]

File: include/hex_quadrature.tcc (hunk 8 above)
Comment (ezrayst) [reply include/hex_quadrature.tcc]: Understood and changed!

**Review by kks32 — APPROVED** [review]

**Pull request merged**